google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/grpc/ChannelPoolHealthChecker.java

```diff
@@ -0,0 +1,253 @@
+  @AutoValue
+  abstract static class ProbeResult {
+    abstract Instant startTime();
+
+    abstract boolean isSuccessful();
+
+    static ProbeResult create(Instant startTime, boolean success) {
+      return new AutoValue_ChannelPoolHealthChecker_ProbeResult(startTime, success);
+    }
+  }
+
+  private final Supplier<ImmutableList<Entry>> entrySupplier;
+  private volatile Instant lastEviction;
+  private final ScheduledExecutorService executor;
+
+  private final ChannelPrimer channelPrimer;
+
+  private ScheduledFuture<?> probeTaskScheduledFuture;
+  private ScheduledFuture<?> detectAndRemoveTaskScheduledFuture;
+
+  private final Clock clock;
+
+  /** Constructor for the pool health checker. */
+  public ChannelPoolHealthChecker(
+      Supplier<ImmutableList<Entry>> entrySupplier,
+      ChannelPrimer channelPrimer,
+      ScheduledExecutorService executor,
+      Clock clock) {
+    this.entrySupplier = entrySupplier;
+    this.lastEviction = Instant.MIN;
+    this.channelPrimer = channelPrimer;
+    this.executor = executor;
+    this.clock = clock;
+  }
+
+  void start() {
+    if (!(channelPrimer instanceof BigtableChannelPrimer)) {
+      logger.log(
+          Level.WARNING,
+          "Provided channelPrimer not an instance of BigtableChannelPrimer, not checking channel health.");
+    } else {
+      Duration initialDelayProbe =
+          Duration.ofMillis(ThreadLocalRandom.current().nextLong(PROBE_INTERVAL.toMillis()));
+      this.probeTaskScheduledFuture =
+          executor.scheduleAtFixedRate(
+              this::runProbes,
+              initialDelayProbe.toMillis(),
+              PROBE_INTERVAL.toMillis(),
+              TimeUnit.MILLISECONDS);
+      Duration initialDelayDetect =
+          Duration.ofMillis(ThreadLocalRandom.current().nextLong(PROBE_INTERVAL.toMillis()));
+      this.detectAndRemoveTaskScheduledFuture =
+          executor.scheduleAtFixedRate(
+              this::detectAndRemoveOutlierEntries,
+              initialDelayDetect.toMillis(),
+              PROBE_INTERVAL.toMillis(),
+              TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /** Stop running health checking */
+  public void stop() {
+    if (probeTaskScheduledFuture != null) {
+      probeTaskScheduledFuture.cancel(false);
+    }
+    if (detectAndRemoveTaskScheduledFuture != null) {
+      detectAndRemoveTaskScheduledFuture.cancel(false);
+    }
+  }
+
+  /** Runs probes on all the channels in the pool. */
+  @VisibleForTesting
+  void runProbes() {
+    Preconditions.checkState(
+        channelPrimer instanceof BigtableChannelPrimer,
+        "Health checking can only be enabled with BigtableChannelPrimer, found %s",
+        channelPrimer);
+    for (Entry entry : this.entrySupplier.get()) {
+      final Instant startTime = clock.instant();
+      final ApiFuture<PingAndWarmResponse> probeFuture;
+
+      BigtableChannelPrimer primer = (BigtableChannelPrimer) channelPrimer;
```

Comment:
this can be a bit easier to read:
- downcast right after the precondition check
- combine declaration and assignment of probeFuture
- inline startTime

```java
Preconditions.checkState(channelPrimer instanceof BigtableChannelPrimer, ...);
BigtableChannelPrimer bigtablePrimer = (BigtableChannelPrimer) channelPrimer;

for (Entry entry : this.entrySupplier.get()) {
  ApiFuture<PingAndWarmResponse> probeFuture = primer.sendPrimeRequestsAsync(entry.getManagedChannel());
  probeFuture.addListener(
          () -> onComplete(entry, clock.instant(), probeFuture), MoreExecutors.directExecutor());
}
```